
# HG changeset patch
# User Magnus Melin <mkmelin+mozilla@iki.fi>
# Date 1752075004 0
# Node ID 45035b144f5ae80613d18f940071082aad389914
# Parent  9d2c2e07e65e3a8e7936f312b4bd11eb1dd1ad70
Bug 1976327 - Don't say address pill ending with with any of .,:;!?- is valid. r=tobyp


When copy-pasting from somehwere, these could easily be found after the address

Differential Revision: https://phabricator.services.mozilla.com/D256651

diff --git a/mail/base/content/widgets/mailWidgets.js b/mail/base/content/widgets/mailWidgets.js
--- a/mail/base/content/widgets/mailWidgets.js
+++ b/mail/base/content/widgets/mailWidgets.js
@@ -996,19 +996,20 @@
         this.updatePill();
       });
     }
 
     /**
      * Simple email address validation.
      *
      * @param {string} address - An email address.
+     * @returns {boolean} true if valid.
      */
     isValidAddress(address) {
-      return /^[^\s@]+@[^\s@]+$/.test(address);
+      return /^[^\s@]+@[^\s@]+[^.,:;!?-]$/.test(address);
     }
 
     /**
      * Convert the pill into "Edit Mode" by hiding the label and showing the
      * html:input element.
      */
     startEditing() {
       // Record the intention of editing a pill as a change in the recipient
@@ -1650,17 +1651,18 @@
       };
     }
 
     /**
      * Create a new recipient pill.
      *
      * @param {HTMLElement} element - The original autocomplete input that
      *   generated the pill.
-     * @param {Array} address - The array containing the recipient's info.
+     * @param {msgIAddressObject[]} address - The array containing the
+     *   recipient's info.
      * @returns {Element} The newly created pill.
      */
     createRecipientPill(element, address) {
       const pill = document.createXULElement("mail-address-pill");
 
       pill.label = address.toString();
       pill.emailAddress = address.email || "";
       pill.fullAddress = address.toString();
diff --git a/mail/test/browser/composition/browser_addressWidgets.js b/mail/test/browser/composition/browser_addressWidgets.js
--- a/mail/test/browser/composition/browser_addressWidgets.js
+++ b/mail/test/browser/composition/browser_addressWidgets.js
@@ -358,16 +358,17 @@ add_task(async function test_pill_creati
   await be_in_folder(accountPOP3.incomingServer.rootFolder);
   const cwc = await open_compose_new_mail();
 
   const addresses = [
     "person@org",
     "foo@address.valid",
     "invalid",
     "foo@address",
+    "bar@example.com.", // trailing dot
   ];
   const subjectField = cwc.document.getElementById("msgSubject");
 
   // Helper method to create multiple pills in a field.
   async function assertPillsCreationInField(input) {
     Assert.ok(input);
     Assert.equal(input.value, "");
 
@@ -445,16 +446,37 @@ add_task(async function test_pill_creati
     );
     // Assert the pill has the correct address.
     Assert.equal(
       input
         .closest(".address-container")
         .querySelectorAll("mail-address-pill")[3].emailAddress,
       addresses[3]
     );
+
+    // Write another invalid address in the field.
+    input.focus();
+    input.value = addresses[4];
+    EventUtils.synthesizeKey("VK_TAB", {}, cwc); // Use Tab to create pill
+
+    // Assert the pill was created.
+    await TestUtils.waitForCondition(
+      () =>
+        input
+          .closest(".address-container")
+          .querySelectorAll("mail-address-pill").length == 5,
+      "Pills created"
+    );
+    // Assert the pill has the correct address.
+    Assert.equal(
+      input
+        .closest(".address-container")
+        .querySelectorAll("mail-address-pill.invalid-address")[1].emailAddress,
+      addresses[4]
+    );
   }
 
   // The To field is visible and focused by default when the compose window is
   // first opened.
   // Test pill creation for the To input field.
   const toInput = cwc.document.getElementById("toAddrInput");
   await assertPillsCreationInField(toInput);
 
@@ -485,17 +507,17 @@ add_task(async function test_pill_creati
     !bccInput.closest(".address-row").classList.contains("hidden"),
     "The Bcc field is visible"
   );
   // Test pill creation for the Bcc input field.
   await assertPillsCreationInField(bccInput);
 
   // Focus on the Bcc field and hold press the Backspace key.
   bccInput.focus();
-  EventUtils.synthesizeKey("KEY_Backspace", { repeat: 5 }, cwc);
+  EventUtils.synthesizeKey("KEY_Backspace", { repeat: 6 }, cwc);
 
   // All pills should be deleted, but the focus should remain on the Bcc field.
   Assert.equal(
     bccInput.closest(".address-container").querySelectorAll("mail-address-pill")
       .length,
     0,
     "All pills in the Bcc field have been removed."
   );
@@ -510,17 +532,17 @@ add_task(async function test_pill_creati
   // Confirm the Bcc field is closed and the focus moved to the Cc field.
   Assert.ok(
     bccInput.closest(".address-row").classList.contains("hidden"),
     "The Bcc field was closed"
   );
   Assert.equal(cwc.document.activeElement, ccInput);
 
   // Now we're on the Cc field. Press and hold Backspace to delete all pills.
-  EventUtils.synthesizeKey("KEY_Backspace", { repeat: 5 }, cwc);
+  EventUtils.synthesizeKey("KEY_Backspace", { repeat: 6 }, cwc);
 
   // All pills should be deleted, but the focus should remain on the Cc field.
   Assert.equal(
     ccInput.closest(".address-container").querySelectorAll("mail-address-pill")
       .length,
     0,
     "All pills in the Cc field have been removed."
   );
@@ -535,17 +557,17 @@ add_task(async function test_pill_creati
   // Confirm the Cc field is closed and the focus moved to the To field.
   Assert.ok(
     ccInput.closest(".address-row").classList.contains("hidden"),
     "The Cc field was closed"
   );
   Assert.equal(cwc.document.activeElement, toInput);
 
   // Now we're on the To field. Press and hold Backspace to delete all pills.
-  EventUtils.synthesizeKey("KEY_Backspace", { repeat: 5 }, cwc);
+  EventUtils.synthesizeKey("KEY_Backspace", { repeat: 6 }, cwc);
 
   // All pills should be deleted, but the focus should remain on the To field.
   Assert.equal(
     toInput.closest(".address-container").querySelectorAll("mail-address-pill")
       .length,
     0,
     "All pills in the To field have been removed."
   );

